### PR TITLE
8283145: [vector] Add the compiler IR and aarch64 support for count leading and trailing zeros

### PR DESCRIPTION
**Pull request by XiaohongGong**
This patch adds the hotspot compiler implementations for the vector operations of count leading and trailing zeros. It adds:
 - two vector nodes respectively for the two vector operations
 - the backend implementations for aarch64 NEON and SVE

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283145](https://bugs.openjdk.java.net/browse/JDK-8283145): [vector] Add the compiler IR and aarch64 support for count leading and trailing zeros

### Reviewers
 * [Ningsheng Jian](https://openjdk.java.net/census#njian) (@nsjian - Committer)
 * [Jatin Bhateja](https://openjdk.java.net/census#jbhateja) (@jatin-bhateja - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-vector pull/180/head:pull/180` \
`$ git checkout pull/180`

Update a local copy of the PR: \
`$ git checkout pull/180` \
`$ git pull https://git.openjdk.java.net/panama-vector pull/180/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 180`

View PR using the GUI difftool: \
`$ git pr show -t 180`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-vector/pull/180.diff">https://git.openjdk.java.net/panama-vector/pull/180.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/aarch64_sve_ad.m4 [hunk 1]
```diff
@@ -3297,3 +3297,52 @@ instruct vmask_gen(pRegGov pg, iRegL len, rFlagsReg cr) %{
...
+  ins_pipe(pipe_slow);
+%}')dnl
+dnl
+dnl
+dnl CLTZ_PREDICATE($1     )
+dnl CLTZ_PREDICATE(op_name)
+define(`CLTZ_PREDICATE', `
+instruct count$1_masked(vReg dst_src, pRegGov pg) %{
+  predicate(UseSVE > 0);
+  match(Set dst_src (Count$1 dst_src pg));
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_sve_ad.m4:3327]: Can you please add comment about why dst and src are the same?

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4 [hunk 2]
```diff
@@ -2513,3 +2513,57 @@ instruct vmask_tolong16B(iRegLNoSp dst, vecX src) %{
...
+    __ umov(rscratch1, as_FloatRegister($ifelse($1, `TrailingZerosV', dst, src)$$reg), __ D, 1);
+    __ clz(rscratch1, rscratch1);
+    __ mov(as_FloatRegister($dst$$reg), __ D, 1, rscratch1);
+  %}
+  ins_pipe(pipe_slow);
+%}')dnl
+dnl
+CLTZ_I(LeadingZerosV, D, 8B/4H/2S)
+CLTZ_I(LeadingZerosV, X, 16B/8H/4S)
+CLTZ_L(LeadingZerosV, 6)
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2564]: Merge these two rules?

**Comment by XiaohongGong** [comment]
Thanks for your review @nsjian ! The comments have been addressed. Thanks!

**Review by nsjian — APPROVED** [review]

**Review by jatin-bhateja — APPROVED** [review]
Common IR changes looks good to me.

**Comment by XiaohongGong** [comment]
Thanks for the review @jatin-bhateja !

**Pull request closed without merging**